packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -60,194 +104,244 @@ export class AssetStaging extends CoreConstruct {
+    // staged asset from the cache.
     this.cacheKey = calculateCacheKey({
+      outdir: this.outdir,
       sourcePath: path.resolve(props.sourcePath),
       bundling: props.bundling,
-      assetHashType: hashType,
+      assetHashType: this.hashType,
+      customFingerprint: this.customSourceFingerprint,
       extraHash: props.extraHash,
       exclude: props.exclude,
+      skip,
     });
 
-    if (props.bundling) {
-      // Check if we actually have to bundle for this stack
-      const bundlingStacks: string[] = this.node.tryGetContext(cxapi.BUNDLING_STACKS) ?? ['*'];
-      const runBundling = !!bundlingStacks.find(pattern => minimatch(Stack.of(this).stackName, pattern));
-      if (runBundling) {
-        const bundling = props.bundling;
-        this.assetHash = AssetStaging.getOrCalcAssetHash(this.cacheKey, () => {
-          // Determine the source hash in advance of bundling if the asset hash type
-          // is SOURCE so that the bundler can opt to re-use its previous output.
-          const sourceHash = hashType === AssetHashType.SOURCE
-            ? this.calculateHash(hashType, props.assetHash, props.bundling)
-            : undefined;
-          this.bundleDir = this.bundle(bundling, outdir, sourceHash);
-          return sourceHash ?? this.calculateHash(hashType, props.assetHash, props.bundling);
-        });
-        this.relativePath = renderAssetFilename(this.assetHash);
-        this.stagedPath = this.relativePath;
-      } else { // Bundling is skipped
-        this.assetHash = AssetStaging.getOrCalcAssetHash(this.cacheKey, () => {
-          return props.assetHashType === AssetHashType.BUNDLE || props.assetHashType === AssetHashType.OUTPUT
-            ? this.calculateHash(AssetHashType.CUSTOM, this.node.path) // Use node path as dummy hash because we're not bundling
-            : this.calculateHash(hashType, props.assetHash);
-        });
-        this.stagedPath = this.sourcePath;
-      }
-    } else {
-      this.assetHash = AssetStaging.getOrCalcAssetHash(this.cacheKey, () => this.calculateHash(hashType, props.assetHash));
-      this.relativePath = renderAssetFilename(this.assetHash, path.extname(this.sourcePath));
-      this.stagedPath = this.relativePath;
-    }
+    const staged = AssetStaging.assetCache.obtain(this.cacheKey, stageThisAsset);
+    this.stagedPath = staged.stagedPath;
+    this.assetHash = staged.assetHash;
+  }
+
+  /**
+   * A cryptographic hash of the asset.
+   *
+   * @deprecated see `assetHash`.
+   */
+  public get sourceHash(): string {
+    return this.assetHash;
+  }
+
+  /**
+   * Return the path to the staged asset, relative to the Cloud Assembly directory of the given stack
+   *
+   * Only returns a relative path if the asset ended up staged inside the outDir,
+   * returns an absolute path if it was not.
+   */
+  public relativeStagedPath(stack: Stack) {
+    const thisAsmDir = Stage.of(stack)?.outdir;
```

Comment:
Why `outdir` here and not `assetOutdir` (or just `this.outdir`)?

Reply:
I'll try to make that more clear through naming.